core/store/src/trie/ops/insert_delete.rs

```diff
@@ -0,0 +1,457 @@
+use near_primitives::errors::StorageError;
+
+use crate::NibbleSlice;
+
+use super::interface::{
+    GenericNodeOrIndex, GenericTrieUpdate, GenericTrieValue, GenericUpdatedNodeId,
+    GenericUpdatedTrieNode, GenericUpdatedTrieNodeWithSize, HasValueLength,
+};
+use super::squash::GenericTrieUpdateSquash;
+
+pub(crate) trait GenericTrieUpdateInsertDelete<'a, N, V>:
+    GenericTrieUpdateSquash<'a, N, V>
+where
+    N: std::fmt::Debug,
+    V: std::fmt::Debug + HasValueLength,
+{
+    fn calc_memory_usage_and_store(
+        &mut self,
+        node_id: GenericUpdatedNodeId,
+        children_memory_usage: u64,
+        new_node: GenericUpdatedTrieNode<N, V>,
+        old_child: Option<GenericUpdatedNodeId>,
+    ) {
+        let new_memory_usage =
+            children_memory_usage.saturating_add(new_node.memory_usage_direct()).saturating_sub(
+                old_child.map(|child| self.get_node_ref(child).memory_usage).unwrap_or_default(),
+            );
+        self.place_node_at(
+            node_id,
+            GenericUpdatedTrieNodeWithSize { node: new_node, memory_usage: new_memory_usage },
+        );
+    }
+
+    /// Insertion logic. We descend from the root down to whatever node corresponds to
+    /// the inserted value. We would need to split, modify, or transform nodes along
+    /// the way to achieve that. This takes care of refcounting changes for existing
+    /// nodes as well as values, but will not yet increment refcount for any newly
+    /// created nodes - that's done at the end.
+    fn generic_insert(
```

Comment:
Personal take: on `GenericTrieUpdateInsertDelete::generic_insert` and similar method we could drop one occurrence of "generic" or even both. 
I'd be ok with a shorter `TrieUpdateInsertDelete::insert`